k9mail/src/main/java/com/fsck/k9/widget/list/MessageListRemoteViewFactory.java

```diff
@@ -0,0 +1,209 @@
+
+        if (cursor == null) {
+            return;
+        }
+
+        try {
+            while (cursor.moveToNext()) {
+                String sender = cursor.getString(0);
+                long date = cursor.isNull(1) ? 0L : cursor.getLong(1);
+                String subject = cursor.getString(2);
+                String preview = cursor.getString(3);
+                boolean unread = toBoolean(cursor.getString(4));
+                boolean hasAttachment = toBoolean(cursor.getString(5));
+                Uri viewUri = Uri.parse(cursor.getString(6));
+
+                mailItems.add(new MailItem(sender, date, subject, preview, unread, hasAttachment, viewUri));
+            }
+        } finally {
+            cursor.close();
+        }
+    }
+
+    @Override
+    public void onDestroy() {
+        // Implement interface
+    }
+
+    @Override
+    public int getCount() {
+        return mailItems.size();
+    }
+
+    @Override
+    public RemoteViews getViewAt(int position) {
+        RemoteViews remoteView = new RemoteViews(context.getPackageName(), R.layout.message_list_widget_list_item);
+        MailItem item = mailItems.get(position);
+
+        CharSequence sender = item.unread ? bold(item.sender) : item.sender;
+        CharSequence subject = item.unread ? bold(item.subject) : item.subject;
+
+        if (senderAboveSubject) {
+            remoteView.setTextViewText(R.id.sender, sender);
+            remoteView.setTextViewText(R.id.mail_subject, subject);
+        } else {
+            remoteView.setTextViewText(R.id.sender, subject);
+            remoteView.setTextViewText(R.id.mail_subject, sender);
+        }
+        remoteView.setTextViewText(R.id.mail_date, item.getDateFormatted("%d %s"));
+        remoteView.setTextViewText(R.id.mail_preview, item.preview);
+
+        int textColor = item.getTextColor();
+        remoteView.setTextColor(R.id.sender, textColor);
+        remoteView.setTextColor(R.id.mail_subject, textColor);
+        remoteView.setTextColor(R.id.mail_date, textColor);
+        remoteView.setTextColor(R.id.mail_preview, textColor);
+
+        if (item.hasAttachment) {
+            remoteView.setInt(R.id.attachment, "setVisibility", View.VISIBLE);
+        } else {
+            remoteView.setInt(R.id.attachment, "setVisibility", View.GONE);
+        }
+
+        Intent intent = new Intent();
+        intent.setData(item.uri);
+        remoteView.setOnClickFillInIntent(R.id.mail_list_item, intent);
+        return remoteView;
+    }
+
+    @Override
+    public RemoteViews getLoadingView() {
+        RemoteViews loadingView = new RemoteViews(context.getPackageName(), R.layout.message_list_widget_loading);
+        loadingView.setTextViewText(R.id.loadingText, context.getString(R.string.mail_list_widget_loading));
+        loadingView.setViewVisibility(R.id.loadingText, View.VISIBLE);
+        return loadingView;
+    }
+
+    @Override
+    public int getViewTypeCount() {
+        return 2;
```

Comment:
For me the method name describes the value well enough.

I don't think the loading view has to be counted extra, but I was too lazy to dig into the Android source code to check. It seems to work fine with the current value.